server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -232,7 +233,8 @@ private static void ensureUnionOutputsHaveCompatibleTypes(AnalyzedRelation left,
         for (int i = 0; i < leftOutputs.size(); i++) {
             var leftType = leftOutputs.get(i).valueType();
             var rightType = rightOutputs.get(i).valueType();
-            if (!DataTypes.isSameType(leftType, rightType)) {
+            if (!DataTypes.isSameType(leftType, rightType) ||
+                leftType.id() == ObjectType.ID && !ObjectType.isCompatibleType((ObjectType) leftType, (ObjectType) rightType)) {
```

Comment:
Would something else break if we changed `isSameType` to do the object type compability check? 

It would have the advantage that we wouldn't have to add objectType special handling in other places.

Reply:
I thought that `isSameType` is reserved for checking exact match. On the other hand, `isCompatibleType` is specialized check for compatibility for unioning.
For example, if one object has `a int` and the other has `b int`, `isSameType` should return `false` but I expect `isCompatibleType` to return `true` as it should be able to be unioned.

Reply:
But what you described is currently not the case and `isSameType` is very lenient. E.g. any two object types will be the same according to the current implementation.

So making it a bit stricter would still fit - and to capture the "but not quite" (e.g. allowing subsets/supersets of object columns). And depending on the current uses of the function we could even rename it.

Reply:
Thanks, I assumed `isSameType` is used everywhere, but now it looks like the main usage is for union and the rest can be adjusted easily.